### PR TITLE
feat(router): add support for custom route reuse strategies

### PR DESCRIPTION
**Pull request by vsavkin**

### CODE REVIEW

File: modules/@angular/router/src/route_reuse_strategy.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+import {ComponentRef} from '@angular/core';
+import {ActivatedRouteSnapshot} from './router_state';
+
+/**
+ * @whatItDoes Represents the detached route tree.
+ *
+ * This is an opaque value the router will give to a custom route reuse strategy
+ * to store and retrieve later on.
```
Comment (vicb) [inline modules/@angular/router/src/route_reuse_strategy.ts]: +1 for store and retrieve (over set/get)

File: modules/@angular/router/src/route_reuse_strategy.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @experimental
+ */
+export type DetachedRouteTree = typeof Object;
+
+/**
+ * @whatItDoes Provides a way to customize when and activated routes get reused.
```
Comment (vicb) [inline modules/@angular/router/src/route_reuse_strategy.ts]: and -> an

File: modules/@angular/router/src/route_reuse_strategy.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @experimental
+ */
+export abstract class RouteReuseStrategy {
+  /**
+   * Determines if this route (and its subtree) should be detached to be reused later.
+   */
+  abstract shouldDettach(route: ActivatedRouteSnapshot): boolean;
```
Comment (vicb) [inline modules/@angular/router/src/route_reuse_strategy.ts]: detach (single t)

File: modules/@angular/router/src/router.ts [hunk 4]
```diff
@@ -1125,6 +1149,22 @@ class ActivateRoutes {
...
+    } else {
+      this.deactiveRouteAndOutlet(route, parentOutletMap);
+    }
+  }
+
+  private detachAndStoreRouteSubtree(
+      route: TreeNode<ActivatedRoute>, parentOutletMap: RouterOutletMap): void {
+    const outlet = getOutlet(parentOutletMap, route.value);
+    const componentRef = outlet.detach();
+    this.routeReuseStrategy.set(route.value.snapshot, <any>{componentRef, route});
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: is `<any>` required here ?

File: modules/@angular/router/src/directives/router_outlet.ts [hunk 5]
```diff
@@ -67,6 +67,13 @@ export class RouterOutlet implements OnDestroy {
     return this._activatedRoute;
   }
 
+  detach(): ComponentRef<any> {
+    this.activated = null;
+    return this.activated;
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_outlet.ts]: needs test (& fix)

File: modules/@angular/router/src/directives/router_outlet.ts [hunk 6]
```diff
@@ -67,6 +67,13 @@ export class RouterOutlet implements OnDestroy {
...
+  }
+
+  reattach(ref: ComponentRef<any>) { this.activated = ref; }
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_outlet.ts]: attach ?

Comment (vicb) [inline modules/@angular/router/src/directives/router_outlet.ts]: Need to update the activated route ?

File: modules/@angular/router/src/router.ts [hunk 7]
```diff
@@ -1087,9 +1104,16 @@ class ActivateRoutes {
       if (future.component) {
         advanceActivatedRoute(future);
         const outlet = getOutlet(parentOutletMap, futureNode.value);
-        const outletMap = new RouterOutletMap();
-        this.placeComponentIntoOutlet(outletMap, future, outlet);
-        this.activateChildRoutes(futureNode, null, outletMap);
+
+        if (this.routeReuseStrategy.shouldAttach(future.snapshot)) {
+          const stored = (<any>this.routeReuseStrategy.get(future.snapshot));
+          outlet.reattach(stored.ref);
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: componentRef ?

File: modules/@angular/router/src/router.ts [hunk 8]
```diff
@@ -1151,6 +1191,11 @@ class ActivateRoutes {
   }
 }
 
+function advancedActivatedRouteNodeAndItsChildren(node: TreeNode<ActivatedRoute>): void {
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: advanceActivatedRouteNodeAndItsChildren (drop the first d)

File: modules/@angular/router/src/create_router_state.ts [hunk 9]
```diff
@@ -8,47 +8,67 @@
...
+    routeReuseStrategy: RouteReuseStrategy, curr: TreeNode<ActivatedRouteSnapshot>,
+    prevState?: TreeNode<ActivatedRoute>): TreeNode<ActivatedRoute> {
+  if (prevState && routeReuseStrategy.shouldReuseRoute(curr.value, prevState.value.snapshot)) {
     const value = prevState.value;
     value._futureSnapshot = curr.value;
-    const children = createOrReuseChildren(curr, prevState);
+    const children = createOrReuseChildren(routeReuseStrategy, curr, prevState);
     return new TreeNode<ActivatedRoute>(value, children);
 
+  } else if (!!routeReuseStrategy.get(curr.value)) {
```
Comment (vicb) [inline modules/@angular/router/src/create_router_state.ts]: - useless `!!`
- not sure to get the logic here

File: modules/@angular/router/src/route_reuse_strategy.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * @whatItDoes Represents the detached route tree.
+ *
+ * This is an opaque value the router will give to a custom route reuse strategy
+ * to store and retrieve later on.
+ *
+ * @experimental
+ */
+export type DetachedRouteTree = typeof Object;
```
Comment (vicb) [inline modules/@angular/router/src/route_reuse_strategy.ts]: finer type ?

File: modules/@angular/router/test/integration.spec.ts [hunk 11]
```diff
@@ -2392,6 +2393,108 @@ describe('Integration', () => {
          })));
     });
   });
+
+  describe('Custom Route Reuse Strategy', () => {
+    class AttacheDetachReuseStrategy implements RouteReuseStrategy {
```
Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: `Attache` no e but may be `reuseStrategy` ?

File: modules/@angular/router/test/integration.spec.ts [hunk 12]
```diff
@@ -2392,6 +2393,108 @@ describe('Integration', () => {
...
+      shouldAttach(route: ActivatedRouteSnapshot): boolean { return false; }
+      get(route: ActivatedRouteSnapshot): DetachedRouteTree { return null; }
+      shouldReuseRoute(future: ActivatedRouteSnapshot, curr: ActivatedRouteSnapshot): boolean {
+        if (future.routeConfig !== curr.routeConfig) {
+          return false;
+        } else if (Object.keys(future.params).length !== Object.keys(curr.params).length) {
+          return false;
+        } else {
+          let equal = true;
+          Object.keys(future.params).forEach(k => {
```
Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: `let equal = Object.keys(future.params).every(...)`

File: modules/@angular/router/test/integration.spec.ts [hunk 13]
```diff
@@ -2392,6 +2393,108 @@ describe('Integration', () => {
...
+            equal = equal && future.params[k] !== curr.params[k];
+          });
+          return equal;
+        }
+      }
+    }
+
+    it('should support attaching & detachmenting fragments',
```
Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: "detachmenting" :)

File: modules/@angular/router/src/route_reuse_strategy.ts (hunk 1 above)
Comment (vsavkin) [inline modules/@angular/router/src/route_reuse_strategy.ts]: Done

File: modules/@angular/router/src/route_reuse_strategy.ts (hunk 2 above)
Comment (vsavkin) [inline modules/@angular/router/src/route_reuse_strategy.ts]: Done

File: modules/@angular/router/src/route_reuse_strategy.ts (hunk 10 above)
Comment (vsavkin) [inline modules/@angular/router/src/route_reuse_strategy.ts]: Introduce the inner type

File: modules/@angular/router/src/router.ts [hunk 14]
```diff
@@ -287,6 +288,20 @@ type NavigationParams = {
   promise: Promise<boolean>
 };
 
+
+/**
+ * Does not detach any subtress. Reuses routes as long as their route config is the same.
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: subtrees

File: modules/@angular/router/src/create_router_state.ts [hunk 15]
```diff
@@ -8,47 +8,67 @@
...
   return new RouterState(root, curr);
 }
 
-function createNode(curr: TreeNode<ActivatedRouteSnapshot>, prevState?: TreeNode<ActivatedRoute>):
-    TreeNode<ActivatedRoute> {
-  if (prevState && equalRouteSnapshots(prevState.value.snapshot, curr.value)) {
+function createNode(
+    routeReuseStrategy: RouteReuseStrategy, curr: TreeNode<ActivatedRouteSnapshot>,
+    prevState?: TreeNode<ActivatedRoute>): TreeNode<ActivatedRoute> {
+  if (prevState && routeReuseStrategy.shouldReuseRoute(curr.value, prevState.value.snapshot)) {
```
Comment (vicb) [inline modules/@angular/router/src/create_router_state.ts]: comment ?

File: modules/@angular/router/src/create_router_state.ts [hunk 16]
```diff
@@ -8,47 +8,67 @@
...
     const value = prevState.value;
     value._futureSnapshot = curr.value;
-    const children = createOrReuseChildren(curr, prevState);
+    const children = createOrReuseChildren(routeReuseStrategy, curr, prevState);
     return new TreeNode<ActivatedRoute>(value, children);
 
+  } else if (routeReuseStrategy.retrieve(curr.value)) {
+    const tree: TreeNode<ActivatedRoute> = (<any>routeReuseStrategy.retrieve(curr.value)).route;
```
Comment (vicb) [inline modules/@angular/router/src/create_router_state.ts]: use the internal type ?

File: modules/@angular/router/src/create_router_state.ts [hunk 17]
```diff
@@ -8,47 +8,67 @@
...
   }
 }
 
+function setFutureSnapshotsOfActivatedRoutes(
+    curr: TreeNode<ActivatedRouteSnapshot>, result: TreeNode<ActivatedRoute>): void {
+  if (curr.value.routeConfig !== result.value.routeConfig) {
+    throw new Error('Cannot reattach ActivatedRouteSnapshot created from a different route');
+  }
+  if (curr.children.length !== result.children.length) {
+    throw new Error('Cannot reattach ActivatedRouteSnapshot with a different number of children');
```
Comment (vicb) [inline modules/@angular/router/src/create_router_state.ts]: not sure about this check ?

File: modules/@angular/router/src/directives/router_outlet.ts [hunk 18]
```diff
@@ -67,11 +67,26 @@ export class RouterOutlet implements OnDestroy {
     return this._activatedRoute;
   }
 
+  detach(): ComponentRef<any> {
+    this.location.detach(0);
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_outlet.ts]: I would not specify the optional `0` (same for `insert`)
Makes me think there are more than 1 view.

File: modules/@angular/router/src/route_reuse_strategy.ts [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * @internal
```
Comment (vicb) [inline modules/@angular/router/src/route_reuse_strategy.ts]: add a comment (on `DetachRouteTree` ?) opaque vs internal

File: modules/@angular/router/src/router.ts [hunk 20]
```diff
@@ -1087,9 +1108,17 @@ class ActivateRoutes {
       if (future.component) {
         advanceActivatedRoute(future);
         const outlet = getOutlet(parentOutletMap, futureNode.value);
-        const outletMap = new RouterOutletMap();
-        this.placeComponentIntoOutlet(outletMap, future, outlet);
-        this.activateChildRoutes(futureNode, null, outletMap);
+
+        if (this.routeReuseStrategy.shouldAttach(future.snapshot)) {
+          const stored: DetachedRouteTreeInternal =
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: can we say `const stored = <DetachedRouteTreeInternal>this.routeReuseStrategy.retrieve(future.snapshot);` ?

File: modules/@angular/router/src/route_reuse_strategy.ts (hunk 10 above)
Comment (vicb) [inline modules/@angular/router/src/route_reuse_strategy.ts]: Could we make it more explicit ?
Would it help with the cast DRT -> DRTI ?
```
export interface DRT {}

export interface DRTI extends DRT {
  // ...
}
```

File: modules/@angular/router/src/create_router_state.ts [hunk 21]
```diff
@@ -8,47 +8,67 @@
...
     const value = prevState.value;
     value._futureSnapshot = curr.value;
-    const children = createOrReuseChildren(curr, prevState);
+    const children = createOrReuseChildren(routeReuseStrategy, curr, prevState);
     return new TreeNode<ActivatedRoute>(value, children);
 
+  } else if (routeReuseStrategy.retrieve(curr.value)) {
+    const tree: TreeNode<ActivatedRoute> = (<any>routeReuseStrategy.retrieve(curr.value)).route;
+    setFutureSnapshotsOfActivatedRoutes(curr, tree);
+    return tree;
```
Comment (vicb) [inline modules/@angular/router/src/create_router_state.ts]: question: should we store `null` at this point ?
/bookmark 1

File: modules/@angular/router/src/router.ts [hunk 22]
```diff
@@ -1125,6 +1154,22 @@ class ActivateRoutes {
 
   private deactiveRouteAndItsChildren(
       route: TreeNode<ActivatedRoute>, parentOutletMap: RouterOutletMap): void {
+    if (this.routeReuseStrategy.shouldDetach(route.value.snapshot)) {
+      this.detachAndStoreRouteSubtree(route, parentOutletMap);
+    } else {
+      this.deactiveRouteAndOutlet(route, parentOutletMap);
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: /ref bookmark 1
or store `null` here so that we never re-use an outdated value ?

File: modules/@angular/router/src/directives/router_outlet.ts (hunk 18 above)
Comment (vicb) [inline modules/@angular/router/src/directives/router_outlet.ts]: if (this.activated) ...

**Comment by vsavkin** [comment]
Got an lgtm from vicb

**Comment by rmcsharry** [comment]
I see that there is a basic blog post about this [here](https://www.softwarearchitekt.at/post/2016/12/02/sticky-routes-in-angular-2-3-with-routereusestrategy.aspx), but it's rather basic and seems to indicate the the reuse strategy is implemented application wide.

Has anyone written a more simple guide than this? Or a way to implement reuse for one specific component only?

**Comment by tjaartvanderWalt** [comment]
@rmcsharry I am also looking for a more detailed post/example. The implementation found [here](https://www.softwarearchitekt.at/post/2016/12/02/sticky-routes-in-angular-2-3-with-routereusestrategy.aspx) breaks when I have paths that contain child paths, I have opened an issue [13869](https://github.com/angular/angular/issues/13869)

**Comment by m98** [comment]
This would be nice if the document could have an example about how to use properly use `RouteReuseStrategy` 

For example, can that be helpful for such situation:

Router:
{path: 'news/:id', component: NewsComponent},  // to read one news
{path: 'news/:id/image/:image-id', component: NewsComponent},  // to see one news image by the id

Now if you click on the image (ex: `news/123/image/5`)  the image opens by the logic we have in the component, but the component also `re-initialize` which means it will rerun the component methods which will retrieve news data over HTTP request.

How can I use `RouteReuseStrategy` for this case?

**Comment by sandangel** [comment]
Don’t know why they miss this in documentation.

**Comment by extrahappyperson** [comment]
Facing the same issue @m98, do they have any updates?
